File: android/app/src/main/java/com/browserselector/android/util/PatternMatcher.kt
package com.browserselector.android.util

import java.net.URI
import java.util.regex.Pattern

/**
 * Utility for matching URLs against wildcard patterns.
 * Supports wildcards:
 * - * matches anything except /
 * - ** matches anything including /
 * - ? matches any single character
 */
object PatternMatcher {

    /**
     * Checks if a URL matches the given pattern.
     */
    fun matches(pattern: String, url: String): Boolean {
        if (pattern.isBlank() || url.isBlank()) return false

        val normalizedUrl = normalizeUrl(url)
        val regex = patternToRegex(pattern)

        return try {
            val compiledPattern = Pattern.compile(regex, Pattern.CASE_INSENSITIVE)
            compiledPattern.matcher(normalizedUrl).matches()
        } catch (e: Exception) {
            false
        }
    }

    /**
     * Converts a wildcard pattern to a regex pattern.
     */
    fun patternToRegex(pattern: String): String {
        val sb = StringBuilder()
        sb.append("^")

        // If pattern doesn't start with scheme, make it optional
        if (!pattern.startsWith("http://") && !pattern.startsWith("https://")) {
            sb.append("(?:https?://)?(?:www\\.)?")
        }

        var i = 0
        while (i < pattern.length) {
            when {
                // Handle ** (matches anything including /)
                i < pattern.length - 1 && pattern[i] == '*' && pattern[i + 1] == '*' -> {
                    sb.append(".*")
                    i += 2
                }
                // Handle * (matches anything except /)
                pattern[i] == '*' -> {
                    sb.append("[^/]*")
                    i++
                }
                // Handle ? (matches any single character)
                pattern[i] == '?' -> {
                    sb.append(".")
                    i++
                }
                // Escape regex special characters
                pattern[i] in "\\[]{}()+^$|" -> {
                    sb.append("\\")
                    sb.append(pattern[i])
                    i++
                }
                // Handle . (escape it)
                pattern[i] == '.' -> {
                    sb.append("\\.")
                    i++
                }
                // Regular character
                else -> {
                    sb.append(pattern[i])
                    i++
                }
            }
        }

        // Allow trailing path/query if not specified
        if (!pattern.contains("/") || pattern.endsWith("*")) {
            sb.append("(?:/.*)?")
        }
        sb.append("$")

        return sb.toString()
    }

    /**
     * Validates a pattern.
     */
    fun isValidPattern(pattern: String): Boolean {
        if (pattern.isBlank()) return false

        return try {
            val regex = patternToRegex(pattern)
            Pattern.compile(regex)
            true
        } catch (e: Exception) {
            false
        }
    }

    /**
     * Extracts the domain from a URL.
     */
    fun extractDomain(url: String): String? {
        return try {
            val normalized = normalizeUrl(url)
            val uri = URI(normalized)
            uri.host?.removePrefix("www.")
        } catch (e: Exception) {
            // Fallback: try regex extraction
            val regex = Regex("(?:https?://)?(?:www\\.)?([^/]+)")
            regex.find(url)?.groupValues?.getOrNull(1)
        }
    }

    /**
     * Converts a domain to a wildcard pattern.
     */
    fun domainToPattern(domain: String): String {
        val cleanDomain = domain.removePrefix("www.").trim()
        return "*.$cleanDomain"
    }

    /**
     * Normalizes a URL by adding https:// if no scheme is present.
     */
    fun normalizeUrl(url: String): String {
        val trimmed = url.trim()
        return when {
            trimmed.startsWith("http://") || trimmed.startsWith("https://") -> trimmed
            else -> "https://$trimmed"
        }
    }

    /**
     * Checks if a URL is valid.
     */
    fun isValidUrl(url: String): Boolean {
        return try {
            val normalized = normalizeUrl(url)
            val uri = URI(normalized)
            uri.scheme != null && uri.host != null
        } catch (e: Exception) {
            false
        }
    }

    /**
     * Truncates a URL for display.
     */
    fun truncateUrl(url: String, maxLength: Int = 60): String {
        if (url.length <= maxLength) return url
        return url.take(maxLength - 3) + "..."
    }
}

File: ios/BrowserSelector/Services/PatternMatcher.swift
import Foundation

/// Utility for matching URLs against wildcard patterns.
/// Supports wildcards:
/// - * matches anything except /
/// - ** matches anything including /
/// - ? matches any single character
enum PatternMatcher {

    /// Checks if a URL matches the given pattern.
    static func matches(pattern: String, url: String) -> Bool {
        guard !pattern.isEmpty, !url.isEmpty else { return false }

        let normalizedUrl = normalizeUrl(url)
        let regex = patternToRegex(pattern)
        let patternLower = pattern.lowercased()
        let domain = extractDomain(from: url)?.lowercased() ?? ""

        do {
            let compiledPattern = try NSRegularExpression(pattern: regex, options: .caseInsensitive)
            let range = NSRange(normalizedUrl.startIndex..., in: normalizedUrl)

            if compiledPattern.firstMatch(in: normalizedUrl, range: range) != nil {
                return true
            }

            // Special handling for plain domain patterns (no wildcards):
            // A pattern like "google.com" should match both "google.com" and "www.google.com"
            if !pattern.contains("*") && !pattern.contains("?") && !pattern.contains("/") {
                if domain.hasSuffix(".\(patternLower)") || domain == patternLower {
                    return true
                }
            }

            // Special handling for *.domain patterns: also match the bare domain
            // Pattern "*.google.com" should also match "google.com"
            if patternLower.hasPrefix("*.") {
                let baseDomain = String(patternLower.dropFirst(2))
                if domain == baseDomain {
                    return true
                }
            }

            return false
        } catch {
            return false
        }
    }

    /// Converts a wildcard pattern to a regex pattern.
    static func patternToRegex(_ pattern: String) -> String {
        var result = "^"

        // If pattern doesn't start with scheme, make it optional
        if !pattern.hasPrefix("http://") && !pattern.hasPrefix("https://") {
            result += "(?:https?://)?(?:www\\.)?"
        }

        var i = pattern.startIndex
        while i < pattern.endIndex {
            let c = pattern[i]

            // Handle ** (matches anything including /)
            if c == "*" && pattern.index(after: i) < pattern.endIndex && pattern[pattern.index(after: i)] == "*" {
                result += ".*"
                i = pattern.index(i, offsetBy: 2)
                continue
            }

            switch c {
            case "*":
                result += "[^/]*"
            case "?":
                result += "."
            case ".", "+", "^", "$", "|", "[", "]", "(", ")", "{", "}", "\\":
                result += "\\\(c)"
            default:
                result += String(c)
            }

            i = pattern.index(after: i)
        }

        // Allow trailing path/query if not specified
        if !pattern.contains("/") || pattern.hasSuffix("*") {
            result += "(?:/.*)?"
        }
        result += "$"

        return result
    }

    /// Validates a pattern.
    static func isValidPattern(_ pattern: String) -> Bool {
        guard !pattern.trimmingCharacters(in: .whitespaces).isEmpty else { return false }

        do {
            let regex = patternToRegex(pattern)
            _ = try NSRegularExpression(pattern: regex)
            return true
        } catch {
            return false
        }
    }

    /// Extracts the domain from a URL.
    static func extractDomain(from url: String) -> String? {
        let normalized = normalizeUrl(url)
        guard let urlObj = URL(string: normalized) else {
            // Fallback: regex extraction
            let regex = try? NSRegularExpression(pattern: "(?:https?://)?(?:www\\.)?([^/]+)")
            let range = NSRange(url.startIndex..., in: url)
            if let match = regex?.firstMatch(in: url, range: range),
               let domainRange = Range(match.range(at: 1), in: url) {
                return String(url[domainRange])
            }
            return nil
        }
        return urlObj.host?.replacingOccurrences(of: "^www\\.", with: "", options: .regularExpression)
    }

    /// Converts a domain to a pattern for rule creation.
    static func domainToPattern(_ domain: String) -> String {
        domain.trimmingCharacters(in: .whitespaces)
            .replacingOccurrences(of: "^www\\.", with: "", options: .regularExpression)
    }

    /// Normalizes a URL by adding https:// if no scheme is present.
    static func normalizeUrl(_ url: String) -> String {
        let trimmed = url.trimmingCharacters(in: .whitespaces)
        if trimmed.hasPrefix("http://") || trimmed.hasPrefix("https://") {
            return trimmed
        }
        return "https://\(trimmed)"
    }

    /// Checks if a URL is valid.
    static func isValidUrl(_ url: String) -> Bool {
        let normalized = normalizeUrl(url)
        guard let urlObj = URL(string: normalized) else { return false }
        return urlObj.scheme != nil && urlObj.host != nil
    }

    /// Truncates a URL for display.
    static func truncateUrl(_ url: String, maxLength: Int = 60) -> String {
        if url.count <= maxLength { return url }
        return String(url.prefix(maxLength - 3)) + "..."
    }
}
